### ISSUE TITLE
distributed.all_gather function stuck when using NCCL backend 

### ISSUE THREAD

**Issue by lecoan** [pytorch/pytorch#18689]
## 🐛 Bug

I am trying to use distributed.all_gather to gather gradients in multi nodes. but I found the all_gather function stuck and no error throw
## To Reproduce

1. set up env variable `MASTER_ADDR` `MASTER_PORT`, save the following code as `main.py`
```python
import torch.distributed as dist
import torch as t

import argparse

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument(
            "--rank",
            type=int,
            help="the rank of proc"
            )
    args = parser.parse_args()
    dist.init_process_group("nccl", rank=args.rank, world_size=2)
    tmp = [t.randn(5).cuda()] * 2
    tensor = t.ones(5).cuda()
    dist.all_gather(tmp, tensor)
    print(tmp)

if __name__ == "__main__":
    main()
```
2. run the code with `python main.py --rank 0` and `python main.py --rank 1`

btw, when I execute this code manually in `ipython`, I found the all_gather quickly go through, but it stuck when trying to print tensor

## Expected behavior

process will print a list of tensor

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```
PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.10.2

Python version: 3.5
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration:
GPU 0: Tesla K80
GPU 1: Tesla K80
GPU 2: Tesla K80
GPU 3: Tesla K80
GPU 4: Tesla K80
GPU 5: Tesla K80
GPU 6: Tesla K80
GPU 7: Tesla K80

Nvidia driver version: 384.145
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.3.1

Versions of relevant libraries:
[pip3] numpy==1.15.3
[pip3] torch==1.0.1.post2
[pip3] torchvision==0.2.1
[conda] Could not collect

**Comment by mrshenli** [comment]
1. Before init the process group, call `torch.cuda.set_device(args.rank)` to assign different GPUs to different processes. 
2. Change `tmp = [t.randn(5).cuda()] * 2` to `tmp = [t.randn(5).cuda() for _ in range(2)]`, otherwise you are using the same tensor to receive two outputs. 

Does this solve your problem?

**Comment by mrshenli** [comment]
NCCL requires inputs to be stored on different GPUs. [link](https://docs.nvidia.com/deeplearning/sdk/nccl-archived/nccl_235/nccl-developer-guide/index.html)

> When creating a communicator, a unique rank between 0 and n-1 has to be assigned to each of the nCUDA devices which are part of the communicator. Using the same CUDA device multiple times as different ranks of the same NCCL communicator is not supported and may lead to hangs.

**Comment by lecoan** [comment]
I change tmp = [t.randn(5).cuda()] * 2 to tmp = [t.randn(5).cuda() for _ in range(2)], and now it works. thank you so much! @mrshenli 
however, if I don't use GPU (eg.  tmp = [t.randn(5)] * 2) and use Gloo, `all_gather` could work. I can't find why

**Comment by mrshenli** [comment]
@lecoan

Do you mean something like the following? It does not hang but still doesn't work for me. The printed tensors are assigned the same value, instead of 0s and 1s, because they are the same tensor. Regarding the hanging behavior, Gloo won't hang even if input tensors are on the same GPU. 

```python
    ws = 2
    t.cuda.set_device(args.rank % 2)
    dist.init_process_group("gloo", rank=args.rank, world_size=ws)
    tmp = [t.randn(5)] * ws
    print (tmp)
    tensor = t.ones(5) * args.rank
    dist.all_gather(tmp, tensor)
    print(tmp)
```

**Comment by mrshenli** [comment]
> use Gloo with [t.randn(5)] * 2 ...

See my response above. If you use your original code, the printed result cannot tell if the output tensors use the same underlying memory space or not. To check that, input tensors on different processes should have different values (i.e., `tensor = t.ones(5) * args.rank`).

**Comment by mrshenli** [comment]
@lecoan may I close this issue?

**Comment by lecoan** [comment]
@mrshenli ok. thanks for your help
